tests/samplers_tests/test_grid.py

```diff
@@ -0,0 +1,165 @@
+    study.optimize(objective, n_trials=n_grids)
+
+    def sorted_values(d):
+        # type: (Dict[str, List[GridValueType]]) -> ValuesView[List[GridValueType]]
+
+        return OrderedDict(sorted(d.items())).values()
+
+    all_grids = itertools.product(*sorted_values(search_space))
+    all_suggested_values = [tuple([p for p in sorted_values(t.params)]) for t in study.trials]
+    assert set(all_grids) == set(all_suggested_values)
+
+    ids = sorted([t.system_attrs['grid_id'] for t in study.trials])
+    assert ids == list(range(n_grids))
+
+    # Test that an optimization fails if the number of trials is more than that of all grids.
+    with pytest.raises(ValueError):
+        study.optimize(objective, n_trials=1)
+
+    # Test a non-existing parameter name in the grid.
+    search_space = {'a': list(range(0, 100, 20))}
+    study = optuna.create_study(sampler=samplers.GridSampler(search_space))
+    with pytest.raises(ValueError):
+        study.optimize(objective)
+
+    # Test a value with out of range.
+    search_space = {
+        'a': [110],  # 110 is out of range specified by the suggest method.
+        'b': [0],
+        'c': ['x'],
+        'd': [0],
+        'e': [0.1]
+    }
+    study = optuna.create_study(sampler=samplers.GridSampler(search_space))
+    with pytest.raises(ValueError):
+        study.optimize(objective)
+
+
+def test_study_optimize_with_multiple_search_spaces():
+    # type: () -> None
+
+    def objective(trial):
+        # type: (Trial) -> float
+
+        a = trial.suggest_int('a', 0, 100)
+        b = trial.suggest_uniform('b', -100, 100)
+
+        return a * b
+
+    # Run 3 trials with a search space.
+    search_space_0 = {
+        'a': [0, 50],
+        'b': [-50, 0, 50]
+    }  # type: Dict[str, List[GridValueType]]
+    sampler_0 = samplers.GridSampler(search_space_0)
+    study = optuna.create_study(sampler=sampler_0)
+    study.optimize(objective, n_trials=3)
+
+    assert len(study.trials) == 3
+    for t in study.trials:
+        sampler_0._same_search_space(t.system_attrs['search_space'])
+
+    # Run 2 trials with another space.
+    search_space_1 = {'a': [0, 25], 'b': [-50]}  # type: Dict[str, List[GridValueType]]
+    sampler_1 = samplers.GridSampler(search_space_1)
+    study.sampler = sampler_1
+    study.optimize(objective, n_trials=2)
+
+    assert not sampler_0._same_search_space(sampler_1._search_space)
+    assert len(study.trials) == 5
+    for t in study.trials[:3]:
+        sampler_0._same_search_space(t.system_attrs['search_space'])
+    for t in study.trials[3: 5]:
+        sampler_1._same_search_space(t.system_attrs['search_space'])
+
+    # Run 3 trials with the first search space again.
+    study.sampler = sampler_0
+    study.optimize(objective, n_trials=3)
+
+    assert len(study.trials) == 8
+    for t in study.trials[:3]:
+        sampler_0._same_search_space(t.system_attrs['search_space'])
+    for t in study.trials[3: 5]:
```

Comment:
[super nit-picking] my taste of style, not mandatory.

```suggestion
    for t in study.trials[3:5]:
```